README.md
# greedy-http

Hello and welcome! `greedy-http` is a dead-simple http interceptor build in `nodeJS` with a minimal set of dependencies.

## What is it?
- A web service to greedily intercept all web requests
- An admin UI to show the details of all web requests
- A websocket service to join the web service and admin UI together

## Quickstart
Although the app is split into 2 halves (api and ui), you can get going with this just by invoking the package directly:

```bash
$ npx greedy-http
```

We also provide a `docker` image courtesy of `ghcr.io` and `docker.io`:

```bash
$ docker run -it -p 3000:3000 -p 3001:3001 -p 3002:3002 thesheps/greedy-http 
```

This will spin the Api, UI and Websocket service up:

```bash
greedy-http api listening on port: 3000...
greedy-http ui listening on port: 3001...
greedy-http websocket listening on port: 3002...
```

## Configuration
`greedy-http` is configured by a few different environment variables:

```bash
GREEDY_HTTP_API_PORT=3000
GREEDY_HTTP_UI_PORT=3001
GREEDY_HTTP_WS_PORT=3002
```

These can be overridden by configuration flags:

```bash
$ npx greedy-http --help

  Usage: greedy http [options] [command]
  
  Commands:
    help     Display help
    version  Display version
  
  Options:
    -A, --api          Enables the api (enabled by default)
    -a, --apiPort <n>  The API port (defaults to 3000)
    -h, --help         Output usage information
    -U, --ui           Enables the admin interface (enabled by default)
    -u, --uiPort <n>   The admin UI port (defaults to 3001)
    -v, --version      Output the version number
    -w, --wsPort <n>   The local websocket port (defaults to 3002)
```

There's a `docker-compose` at the root of this repo that you can also use as a reference, if that bakes your noodle!

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.5.0 (2022-11-10)


### Features

* adds a dead basic docker-compose ([a905b79](https://github.com/thesheps/greedy-http/commit/a905b7926c59002cb9913364976cd42c94950486))
* adds dead basic Dockerfile ([97cdfb7](https://github.com/thesheps/greedy-http/commit/97cdfb728de992eb94ea240282523c76a2142225))
* adds headers component ([3741d46](https://github.com/thesheps/greedy-http/commit/3741d46769d9bb74ff08d4d72ede8924a3135ed0))
* adds request lister component ([d7d5804](https://github.com/thesheps/greedy-http/commit/d7d5804a8a2cbb6793de3c5622cb0ff81f5af64d))
* **all:** removes docker as dependency, introduces bin entrypoint ([098591f](https://github.com/thesheps/greedy-http/commit/098591fd015509644147c9d3ccdda29d61904dd2))
* **api:** adds barebones api with redis, webdis and docker-compose ([da88dce](https://github.com/thesheps/greedy-http/commit/da88dce3355a9a973cf9e6f55b940a8b76e302b8))
* **api:** removes unneeded redis dependencies, adds simple websocket implementation ([c17c000](https://github.com/thesheps/greedy-http/commit/c17c000dedb7d7cb8f0992f953da6fc0d34323de))
* **ci:** Enables push to ghcr ([f03df9f](https://github.com/thesheps/greedy-http/commit/f03df9fc590b76410f3925f2b3cdfa1b13af2035))
* **ci:** Enables push to ghcr ([2239c4d](https://github.com/thesheps/greedy-http/commit/2239c4da8cf961cd11e0e9e0318c01295f73efa4))
* **ci:** Enables push to ghcr ([a2bdb3c](https://github.com/thesheps/greedy-http/commit/a2bdb3cb09d35244ac9900125cd83fa8e50fab36))
* **Dockerfile:** actually installs dependencies ([26957e1](https://github.com/thesheps/greedy-http/commit/26957e194b9603bf2f170fbda4fec13518c0d3cc))
* parameterises api and ui run functions ([a4a9730](https://github.com/thesheps/greedy-http/commit/a4a97305e2d7a56b58ee2b141348c4c81e2775f9))
* **server:** initial check-in ([1ccc303](https://github.com/thesheps/greedy-http/commit/1ccc3039664c8c8e1612ecbd1ba1ebb98f23022a))
* **tool:** adds command line arg handling ([dd30af5](https://github.com/thesheps/greedy-http/commit/dd30af5dff9962adc9f9dd2c44ff248797880e22))
* **ui:** adds barebones ui to docker-compose ([aa238ec](https://github.com/thesheps/greedy-http/commit/aa238ec5f9baca3e724cc3263ce2bf12c54c4a74))
* **ui:** adds body renderer ([f534449](https://github.com/thesheps/greedy-http/commit/f534449ff4a13b3d5cf7d249b27677e34d8d48d2))
* **ui:** adds item selection styling ([d0c20fc](https://github.com/thesheps/greedy-http/commit/d0c20fc82942ee7cca0ad267ebe179d5775df0f1))
* **ui:** adds login flow ([1820674](https://github.com/thesheps/greedy-http/commit/1820674037fa47817088b2891ba533cb144ef54f))
* **ui:** adds navbar webcomponent ([618cee3](https://github.com/thesheps/greedy-http/commit/618cee3653872a0d6799ff1cabcfd67214ee7c06))
* **ui:** adds top-level app component ([0f32483](https://github.com/thesheps/greedy-http/commit/0f32483d26d3e6c9495ab34ae18e94dcc62f9c7a))
* **ui:** implements simple store ([36c2353](https://github.com/thesheps/greedy-http/commit/36c2353add8103626d89f70c6b2b5891e7cee8f6))


### Bug Fixes

* **entrypoint:** fixes reference from index.js ([f91b863](https://github.com/thesheps/greedy-http/commit/f91b863f64ef53259ce3fa88a4437c3344ffc5d2))

### [1.4.1](https://github.com/thesheps/greedy-http/compare/v1.4.0...v1.4.1) (2022-11-10)

## [1.4.0](https://github.com/thesheps/greedy-http/compare/v1.3.5...v1.4.0) (2022-11-10)


### Features

* **ui:** adds title to UI ([21c2f3e](https://github.com/thesheps/greedy-http/commit/21c2f3e408a3e906c36b8c579f635918d788ea95))

### [1.3.5](https://github.com/thesheps/greedy-http/compare/v1.1.0...v1.3.5) (2022-11-10)


server.js
const http = require("http");
const static = require("node-static");
const ws = require("ws");

const apiPort = process.env.API_PORT ?? 3000;
const uiPort = process.env.UI_PORT ?? 3001;
const wsPort = process.env.WEBSOCKET_PORT ?? 3002;

function run() {
  const wsServer = new ws.WebSocketServer({ port: wsPort });
  const staticServer = new static.Server(`${__dirname}/ui`);

  http.createServer((req, res) => staticServer.serve(req, res)).listen(uiPort);

  http
    .createServer(async (req, res) => {
      const payload = JSON.stringify({
        timestamp: new Date().getTime(),
        remoteAddress: req.socket.remoteAddress,
        headers: req.headers,
      });

      wsServer.clients.forEach((client) => client.send(payload));

      res.statusCode = 200;
      res.end();
    })
    .listen(apiPort);
}

module.exports = {
  run,
};
